File: MuGeonGiV2/Core/EchoEffector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using CSCore;
using CSCore.DMO.Effects;
using CSCore.Streams.Effects;
using Newtonsoft.Json;

namespace MuGeonGiV2.Core
{
    public class EchoEffector : Effector
    {
        [JsonProperty]
        public float RightDelay
        {
            get => _rightDelay;
            set
            {
                if (_effect != null)
                {
                    _effect.RightDelay = value;
                }
                _rightDelay = value;
            }
        }

        [JsonProperty]
        public float WetDryMix
        {
            get => _wetDryMix;
            set
            {
                if (_effect != null)
                {
                    _effect.WetDryMix = value;
                }
                _wetDryMix = value;
            }
        }
        [JsonProperty]
        public float Feedback
        {
            get => _feedback;
            set
            {
                if (_effect != null)
                {
                    _effect.Feedback = value;
                }
                _feedback = value;
            }
        }
        [JsonProperty]
        public float LeftDelay
        {
            get => _leftDelay;
            set
            {
                if (_effect != null)
                {
                    _effect.LeftDelay = value;
                }
                _leftDelay = value;
            }
        }
        [JsonProperty]
        public int PanDelay
        {
            get => _panDelay;
            set
            {
                if (_effect != null)
                {
                    _effect.PanDelay = value == 1;
                }
                _panDelay = value;
            }
        }

        private float _rightDelay;
        private float _wetDryMix;
        private float _feedback;
        private float _leftDelay;
        private int _panDelay;

        [JsonProperty]
        public const float WetDryMixMin = 0;
        [JsonProperty]
        public const float WetDryMixMax = 100;
        [JsonProperty]
        public const float RightDelayMin = 1;
        [JsonProperty]
        public const float RightDelayMax = 2000;
        [JsonProperty]
        public const float LeftDelayMin = 1;
        [JsonProperty]
        public const float LeftDelayMax = 2000;
        [JsonProperty]
        public const float FeedbackMin = 0;
        [JsonProperty]
        public const float FeedbackMax = 100;
        [JsonProperty]
        public const int PanDelayMax = 1;
        [JsonProperty]
        public const int PanDelayMin = 0;

        private DmoEchoEffect _effect;

        public EchoEffector()
        {
            RightDelay = DmoEchoEffect.RightDelayDefault;
            WetDryMix = DmoEchoEffect.WetDryMixDefault;
            Feedback = DmoEchoEffect.FeedbackDefault;
            LeftDelay = DmoEchoEffect.LeftDelayDefault;
            PanDelay = DmoEchoEffect.PanDelayDefault ? 1 : 0;
        }
        public override IWaveSource AppendSource(IWaveSource source)
        {
            _effect = new DmoEchoEffect(source)
            {
                RightDelay = RightDelay,
                WetDryMix = WetDryMix,
                Feedback = Feedback,
                LeftDelay = LeftDelay,
                PanDelay = (PanDelay == 1),
            };

            return _effect;
        }
    }
}

File: ui/src/instruments/EchoEffector.js
import React from 'react';
import Instrument from './Instrument';
import SingleBox from './SingleBox';
import SettingPortal from './SettingPortal';

export default class EchoEffector extends Instrument {
  static StatesWillSave = [
    'rightDelay',
    'wetDryMix',
    'feedback',
    'leftDelay',
    'panDelay',
    'rightDelayMin',
    'rightDelayMax',
    'wetDryMixMin',
    'wetDryMixMax',
    'feedbackMin',
    'feedbackMax',
    'leftDelayMin',
    'leftDelayMax',
    'panDelayMin',
    'panDelayMax',
  ];
  // 1. min max 값 가져와서 값 수정할 때 min, max 조절해주고
  // 2. setter 만들어서 react state 바꿔주고, 서버에다가도 바꿔주기
  render() {
    console.log(this.props);
    console.log(this.state);
    return (
      <SingleBox {...this.props} >
        Echo Effector

        <SettingPortal
          {...this.props}
        >
          {'<Echo Effector>'}
          <h3>rightDelay</h3>
          <input
            type="range"
            value={this.state.rightDelay}
            min={this.state.rightDelayMin}
            max={this.state.rightDelayMax}
            onChange={(event) => { this.rightDelay = event.target.value; }}
          />
          <span>{this.state.rightDelay}</span>
          <h3>wetDryMix</h3>
          <input
            type="range"
            value={this.state.wetDryMix}
            min={this.state.wetDryMixMin}
            max={this.state.wetDryMixMax}
            onChange={(event) => { this.wetDryMix = event.target.value; }}
          />
          <span>{this.state.wetDryMix}</span>
          <h3>feedback</h3>
          <input
            type="range"
            value={this.state.feedback}
            min={this.state.feedbackMin}
            max={this.state.feedbackMax}
            onChange={(event) => { this.feedback = event.target.value; }}
          />
          <span>{this.state.feedback}</span>
          <h3>leftDelay</h3>
          <input
            type="range"
            value={this.state.leftDelay}
            min={this.state.leftDelayMin}
            max={this.state.leftDelayMax}
            onChange={(event) => { this.leftDelay = event.target.value; }}
          />
          <span>{this.state.leftDelay}</span>
          <h3>panDelay</h3>
          <input
            type="range"
            value={this.state.panDelay}
            min={this.state.panDelayMin}
            max={this.state.panDelayMax}
            onChange={(event) => { this.panDelay = event.target.value; }}
          />
          <span>{this.state.panDelay}</span>
        </SettingPortal>
      </SingleBox>
    );
  }
}
